test/unit/test_bech32.c

```diff
@@ -0,0 +1,218 @@
+	{ "tb1qqqqqp399et2xygdj5xreqhjjvcmzhxw4aywxecjdzew6hylgvsesrxh6hy",
+		34, { 0x00, 0x20, 0x00, 0x00, 0x00, 0xc4, 0xa5, 0xca, 0xd4, 0x62, 0x21, 0xb2, 0xa1, 0x87, 0x90, 0x5e, 0x52, 0x66, 0x36, 0x2b, 0x99, 0xd5, 0xe9, 0x1c, 0x6c, 0xe2, 0x4d, 0x16, 0x5d, 0xab, 0x93, 0xe8, 0x64, 0x33 } },
+	{ "tb1pqqqqp399et2xygdj5xreqhjjvcmzhxw4aywxecjdzew6hylgvsesf3hn0c",
+		34, { 0x51, 0x20, 0x00, 0x00, 0x00, 0xc4, 0xa5, 0xca, 0xd4, 0x62, 0x21, 0xb2, 0xa1, 0x87, 0x90, 0x5e, 0x52, 0x66, 0x36, 0x2b, 0x99, 0xd5, 0xe9, 0x1c, 0x6c, 0xe2, 0x4d, 0x16, 0x5d, 0xab, 0x93, 0xe8, 0x64, 0x33 } },
+	{ "bc1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vqzk5jj0",
+		34, { 0x51, 0x20, 0x79, 0xbe, 0x66, 0x7e, 0xf9, 0xdc, 0xbb, 0xac, 0x55, 0xa0, 0x62, 0x95, 0xce, 0x87, 0x0b, 0x07, 0x02, 0x9b, 0xfc, 0xdb, 0x2d, 0xce, 0x28, 0xd9, 0x59, 0xf2, 0x81, 0x5b, 0x16, 0xf8, 0x17, 0x98 } },
+};
+
+static const char *invalid_address[] = {
+	"tc1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vq5zuyut",
+	"bc1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vqh2y7hd",
+	"tb1z0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vqglt7rf",
+	"BC1S0XLXVLHEMJA6C4DQV22UAPCTQUPFHLXM9H8Z3K2E72Q4K9HCZ7VQ54WELL",
+	"bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kemeawh",
+	"tb1q0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vq24jc47",
+	"bc1p38j9r5y49hruaue7wxjce0updqjuyyx0kh56v8s25huc6995vvpql3jow4",
+	"BC130XLXVLHEMJA6C4DQV22UAPCTQUPFHLXM9H8Z3K2E72Q4K9HCZ7VQ7ZWS8R",
+	"bc1pw5dgrnzv",
+	"bc1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7v8n0nx0muaewav253zgeav",
+	"BC1QR508D6QEJXTDG4Y5R3ZARVARYV98GJ9P",
+	"tb1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vq47Zagq",
+	"bc1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7v07qwwzcrf",
+	"tb1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vpggkg4j",
+	"bc1gmk9yu",
+};
+
+static struct invalid_address_data invalid_address_enc[] = {
+	{ "BC", 0, 20 },
+	{ "bc", 0, 21 },
+	{ "bc", 17, 32 },
+	{ "bc", 1, 1 },
+	{ "bc", 16, 41 },
+};
+
+static void segwit_scriptpubkey(uint8_t *scriptpubkey, size_t *scriptpubkeylen, int witver, const uint8_t *witprog, size_t witprog_len) {
+	scriptpubkey[0] = witver? (0x50 + witver): 0;
+	scriptpubkey[1] = witprog_len;
+	memcpy (scriptpubkey + 2, witprog, witprog_len);
+	*scriptpubkeylen = witprog_len + 2;
+}
+
+int my_strncasecmp (const char *s1, const char *s2, size_t n) {
+	size_t i = 0;
+	while (i < n) {
+		char c1 = s1[i];
+		char c2 = s2[i];
+		if (c1 >= 'A' && c1 <= 'Z')
+			c1 = (c1 - 'A') + 'a';
+		if (c2 >= 'A' && c2 <= 'Z')
+			c2 = (c2 - 'A') + 'a';
+		if (c1 < c2)
+			return -1;
+		if (c1 > c2)
+			return 1;
+		if (c1 == 0)
+			return 0;
+		++i;
+	}
+	return 0;
+}
+
+void test_crypto_bech32_encode (void) {
+	uint8_t data[82];
+	char rebuild[92];
+	char hrp[84];
+	size_t data_len;
+	int i;
+	for (i = 0; i < sizeof (valid_checksum_bech32) / sizeof (valid_checksum_bech32[0]); i++) {
+		bech32_decode (hrp, data, &data_len, valid_checksum_bech32[i]);
```

Comment:
these tests use an api that shouldnt be made public to protect the abi contract. so you can:

* use RCrypto api to test bech32 with its plugin instead 
* expose this API via R2_USE_NEW_ABI, and prefix it with r_crypto_bech32_...